xpallets/gateway/bitcoin/v2/src/redeem.rs

```diff
@@ -0,0 +1,285 @@
+    #[pallet::getter(fn redeem_fee)]
+    pub(crate) type RedeemFee<T: Config> = StorageValue<_, u8, ValueQuery>;
+
+    /// Auto-increament id to identity each request
+    #[pallet::storage]
+    pub(crate) type RedeemId<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+    /// Mapping from redeem id to `RedeemRequest`
+    #[pallet::storage]
+    pub(crate) type RedeemRequests<T: Config> =
+        StorageMap<_, Twox64Concat, super::types::RedeemRequestIdentify, RedeemRequest<T>>;
+
+    /// Expired time for an `RedeemRequest`
+    #[pallet::storage]
+    pub(crate) type RedeemRequestExpiredTime<T: Config> =
+        StorageValue<_, BlockNumberFor<T>, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// User request redeem
+        #[pallet::weight(0)]
+        pub fn request_redeem(
+            origin: OriginFor<T>,
+            vault_id: T::AccountId,
+            redeem_amount: BalanceOf<T>,
+            btc_addr: super::types::BtcAddress,
+        ) -> DispatchResultWithPostInfo {
+            let sender = ensure_signed(origin)?;
+            let height = <frame_system::Pallet<T>>::block_number();
+            let vault = vaultpallet::Pallet::<T>::get_active_vault_by_id(&vault_id)?;
+
+            //check current bridge status
+            let bridge_status = <BridgeStatus<T>>::get();
+            match bridge_status {
+                crate::assets::types::Status::Running => {}
+                crate::assets::types::Status::Error => {
+                    Self::deposit_event(Event::<T>::ChainStatusErro);
+                    return Ok(().into());
+                }
+                crate::assets::types::Status::Shutdown => {
+                    Self::deposit_event(Event::<T>::ChainStatusErro);
+                    return Ok(().into());
+                }
+            }
+
+            // decrase user's XBTC amount
+
+            // decrese vault's lock collateral.
+            let worth_pcx = assetspallet::Pallet::<T>::convert_to_pcx(redeem_amount)?;
+            assetspallet::Pallet::<T>::lock_collateral(&vault_id, worth_pcx)?;
+
+            // tell vault to transfer btc to this btc_addr
+
+            // generate redeem request identify
+            let redeemid: super::types::RedeemRequestIdentify = Vec::new();
+            let redeemidentify: super::types::RedeemRequestIdentify = redeemid;
+
+            // insert RedeemRequest into map
+            <RedeemRequests<T>>::insert(
+                redeemidentify,
+                RedeemRequest::<T> {
+                    vault: vault_id,
+                    open_time: height,
+                    requester: sender,
+                    btc_address: btc_addr,
+                    amount: redeem_amount,
+                    redeem_fee: Default::default(),
+                    status: super::types::RedeemRequestStatus::WaitForGetBtc,
+                },
+            );
+
+            // Redeem id increase
+            <RedeemId<T>>::mutate(|c| *c += 1);
```

Comment:
replace the fake redeemid mentioned with this.

Reply:
ok. got it